File: policy-agent/internal/agent/orchestrator.go
package agent

import (
	"context"
	"fmt"
	"time"

	"policy-agent/internal/ai"
	"policy-agent/internal/policy"
	"policy-agent/internal/validator"
	"policy-agent/pkg/errors"
	"policy-agent/pkg/types"
)

// Orchestrator coordinates validation across multiple domains
type Orchestrator struct {
	registry validator.Registry
	engine   *policy.Engine
	aiClient ai.Client
}

// New creates a new orchestrator
func New(registry validator.Registry, engine *policy.Engine, aiClient ai.Client) *Orchestrator {
	return &Orchestrator{
		registry: registry,
		engine:   engine,
		aiClient: aiClient,
	}
}

// Validate performs validation on the given input
func (o *Orchestrator) Validate(ctx context.Context, req *ValidateRequest) (*ValidateResponse, error) {
	startTime := time.Now()

	// Determine which validator to use
	var v validator.Validator
	var err error

	if req.Domain != "" {
		// Use domain-specific validator
		v, err = o.registry.Get(req.Domain)
		if err != nil {
			return nil, err
		}
	} else if req.ResourceType != "" {
		// Use resource type to determine validator
		v, err = o.registry.GetForResourceType(req.ResourceType)
		if err != nil {
			return nil, err
		}
	} else {
		return nil, errors.New(errors.ErrCodeConfigInvalid, "either domain or resource_type must be specified")
	}

	// Prepare validator input
	input := &validator.Input{
		Raw:          req.Raw,
		Format:       req.Format,
		ResourceType: req.ResourceType,
		Metadata:     req.Metadata,
		File:         req.File,
	}

	// Validate
	result, err := v.Validate(ctx, input)
	if err != nil {
		return nil, errors.Wrap(err, errors.ErrCodeValidationFailed, "validation failed")
	}

	result.Duration = time.Since(startTime)

	// If auto-fix is requested and there are violations, attempt remediation
	if req.AutoFix && len(result.Violations) > 0 && o.aiClient != nil {
		fixReq := &ai.RemediateRequest{
			Domain:         result.Domain,
			OriginalConfig: string(req.Raw),
			Violations:     result.Violations,
		}

		fixResp, err := o.aiClient.Remediate(ctx, fixReq)
		if err == nil && fixResp != nil {
			// Add AI remediation suggestions to violations
			for i := range result.Violations {
				if i < len(fixResp.Suggestions) {
					result.Violations[i].Remediation = &types.Remediation{
						AutoFixable:   fixResp.Suggestions[i].AutoFixable,
						Suggestion:    fixResp.Suggestions[i].Suggestion,
						AIExplanation: fixResp.Suggestions[i].Explanation,
					}
				}
			}
		}
	}

	return &ValidateResponse{
		Result: result,
	}, nil
}

// Generate generates a policy-compliant configuration using AI
func (o *Orchestrator) Generate(ctx context.Context, req *types.GenerateRequest) (*types.GenerateResponse, error) {
	if o.aiClient == nil {
		return nil, errors.New(errors.ErrCodeConfigInvalid, "AI client not configured")
	}

	// Convert to AI request
	aiReq := &ai.GenerateRequest{
		Domain:       req.Domain,
		Requirements: req.Requirements,
		Context:      req.Context,
		Policies:     req.Policies,
	}

	// Call AI to generate configuration
	resp, err := o.aiClient.Generate(ctx, aiReq)
	if err != nil {
		return nil, errors.Wrap(err, errors.ErrCodeAIRequestFailed, "AI generation failed")
	}

	return &types.GenerateResponse{
		Configuration: resp.Configuration,
		Format:        resp.Format,
		Explanation:   resp.Explanation,
		PoliciesMet:   resp.PoliciesMet,
	}, nil
}

// Fix attempts to fix policy violations in a configuration
func (o *Orchestrator) Fix(ctx context.Context, req *types.FixRequest) (*types.FixResponse, error) {
	if o.aiClient == nil {
		return nil, errors.New(errors.ErrCodeConfigInvalid, "AI client not configured")
	}

	// First, validate to find violations
	validateReq := &ValidateRequest{
		Raw:    []byte(req.File),
		Format: types.FormatYAML, // TODO: Detect format
	}

	validateResp, err := o.Validate(ctx, validateReq)
	if err != nil {
		return nil, err
	}

	// If no violations, nothing to fix
	if len(validateResp.Result.Violations) == 0 {
		return &types.FixResponse{
			Original: req.File,
			Fixed:    req.File,
			Changes:  []types.Change{},
			Explanation: "No policy violations found",
		}, nil
	}

	// Use AI to remediate
	remediateReq := &ai.RemediateRequest{
		Domain:         validateResp.Result.Domain,
		OriginalConfig: req.File,
		Violations:     validateResp.Result.Violations,
	}

	remediateResp, err := o.aiClient.Remediate(ctx, remediateReq)
	if err != nil {
		return nil, errors.Wrap(err, errors.ErrCodeAIRequestFailed, "remediation failed")
	}

	// Convert AI suggestions to changes
	changes := make([]types.Change, len(remediateResp.Suggestions))
	for i, suggestion := range remediateResp.Suggestions {
		changes[i] = types.Change{
			Field:    validateResp.Result.Violations[i].Field,
			OldValue: validateResp.Result.Violations[i].CurrentValue,
			NewValue: validateResp.Result.Violations[i].ExpectedValue,
			Reason:   suggestion.Suggestion,
		}
	}

	return &types.FixResponse{
		Original:    req.File,
		Fixed:       remediateResp.FixedConfig,
		Changes:     changes,
		Explanation: remediateResp.Explanation,
	}, nil
}

// ValidateRequest represents a validation request
type ValidateRequest struct {
	Raw          []byte
	Format       types.Format
	Domain       string
	ResourceType string
	Metadata     map[string]interface{}
	File         string
	AutoFix      bool
}

// ValidateResponse represents a validation response
type ValidateResponse struct {
	Result *types.ValidationResult
}

// GetAvailableDomains returns all available validation domains
func (o *Orchestrator) GetAvailableDomains() []string {
	return o.registry.Domains()
}

// GetPolicyPackages returns all loaded policy packages
func (o *Orchestrator) GetPolicyPackages() []string {
	return o.engine.GetPolicyPackages()
}

File: policy-agent-py/policy_agent/agent/orchestrator.py
"""Orchestrator for coordinating validation across domains."""

from datetime import datetime
from typing import Any, Dict, List, Optional

import yaml

from policy_agent.validators.base import Validator, ValidatorRegistry
from policy_agent.policy.engine import PolicyEngine
from policy_agent.types.result import (
    ValidationResult,
    GenerateRequest,
    GenerateResponse,
    FixRequest,
    FixResponse,
)


class Orchestrator:
    """Coordinates validation across multiple domains."""

    def __init__(
        self,
        validators: Optional[List[Validator]] = None,
        policy_engine: Optional[PolicyEngine] = None,
        ai_client: Optional[Any] = None,
    ):
        """Initialize orchestrator.

        Args:
            validators: List of validators to register
            policy_engine: Policy engine instance
            ai_client: AI client for generation/remediation
        """
        self.registry = ValidatorRegistry()
        self.policy_engine = policy_engine
        self.ai_client = ai_client

        if validators:
            for validator in validators:
                self.registry.register(validator)

    def validate(
        self,
        content: str,
        domain: Optional[str] = None,
        file_path: Optional[str] = None,
    ) -> ValidationResult:
        """Validate configuration content.

        Args:
            content: YAML/JSON configuration content
            domain: Specific domain to validate (optional)
            file_path: Source file path (optional)

        Returns:
            ValidationResult with violations and warnings
        """
        start_time = datetime.now()

        # Parse YAML
        try:
            data = yaml.safe_load(content)
        except yaml.YAMLError as e:
            raise ValueError(f"Failed to parse YAML: {e}")

        # Determine validator
        if domain:
            validator = self.registry.get(domain)
        else:
            # Auto-detect from resource kind
            kind = data.get("kind", "")
            if not kind:
                raise ValueError("Resource kind not specified and domain not provided")
            validator = self.registry.get_for_resource_type(kind)

        # Validate
        result = validator.validate(data, file_path)

        # Calculate duration
        duration = (datetime.now() - start_time).total_seconds()
        result.duration = duration

        return result

    def generate(self, request: GenerateRequest) -> GenerateResponse:
        """Generate policy-compliant configuration using AI.

        Args:
            request: Generation request

        Returns:
            GenerateResponse with generated configuration

        Raises:
            ValueError: If AI client not configured
        """
        if not self.ai_client:
            raise ValueError("AI client not configured")

        return self.ai_client.generate(request)

    def fix(self, request: FixRequest) -> FixResponse:
        """Fix policy violations using AI.

        Args:
            request: Fix request

        Returns:
            FixResponse with fixed configuration

        Raises:
            ValueError: If AI client not configured
        """
        if not self.ai_client:
            raise ValueError("AI client not configured")

        # First validate to find violations
        with open(request.file, "r") as f:
            content = f.read()

        result = self.validate(content, file_path=request.file)

        if not result.violations:
            # No violations, nothing to fix
            return FixResponse(
                original=content,
                fixed=content,
                changes=[],
                explanation="No policy violations found",
            )

        # Use AI to fix
        return self.ai_client.remediate(result.domain, content, result.violations)

    def get_available_domains(self) -> List[str]:
        """Get list of available validation domains.

        Returns:
            List of domain names
        """
        return self.registry.domains()
